20210530-acme-zhiwei.md

```diff
@@ -0,0 +1,175 @@
+
+1. Master: In a cluster, a single DNS Server can be configured as "master", which would be responsible for actually performing ACME challenges and obtaining certificate from the Let's Encrypt CA. The Master would then store the certificate in the zone file as CERT RR under the domain of `dns.coredns.io`. This certificate can be used to serve DoH whenever there is a DNS query to the master.
+
+2. Worker: The worker is responsible for fetching the certificates from the master. The worker is not responsible for performing ACME challenges. Whenever a DNS query is received at the worker node, the worker first queries the master for the certificate. Workers should cache the certificate and use it for further DoH queries.
+
+Configurations:
+
+Master
+```
+https://example.com {
+  tls {
+    master
+    acme <DOMAIN> {
+      challenge <challenge_type>
+      /* challenge spec */
+    }
+  }
+}
+```
+
+Worker
+```
+https://example.com {
+  tls {
+    worker
+  }
+  cache
+}
+```
+
+The worker caches the CERT RR (for X seconds) to reduce querying the master for the certificate every time a DNS query is received.
+
+#### Cert Renewal in Clustered Environment
+
+In order to prevent the workers from using the stale certificates, the master disseminates the renewed certificate to them instead of workers polling the master for the certificates.
+
+## Timeline
+Due to the tight timeline, 50% of the project has to be implemented by July 11th.
+Period: June 1st - August 31st
+
+**31 May - 6 June**
+1. Approval of RFC
+2. Setting up of development environment
+3. First Meeting
+4. After approval of RFC, start of writing
+
+**7 June - 20 June**
+1. Develop code to perform ACME, its basic functionality, along with documentation and tests
+
+**21 June - 5 July**
+1. Develop code to handle HTTP01 and DNS01 Challenges, along with documentation and tests
+
+**6 July - 25 July**
+50% implementation done by 11 July
+1. Initial implementation of master and workers
+2. Creation of the clustered environment for development and testing
+
+**26 July -  15 August**
+1. Implementation of certificate renewal
+2. More rigorous testing of clustered implementation
+
+**16 August - 31 August**
+1. Buffer period
+2. Cleanup
+3. Finish writing the final report
+
+## Questions and Discussion Topics
+
+1. How should I configure the workers to look for the master? When the certificate is renewed, how does the master know where the workers are and how to send the certificate to them? I am aware that this is an issue later on when we implement the clustered environment and hence not urgent yet.
+2. Where should I store the certificate and public-private key pair?
+3. Any idea about certificate renewal? The previous RFC mentioned changing the `reload` plugin to watch the expiration of the CERT RR.
+4. Should I create a new plugin or modify the tls plugin? I would recommend modifying the TLS plugin instead since the parser for `tls` keyword exists inside the `tls` plugin.
```

Comment:
I think ideally we want to have a new plugin, as coredns has the philosophy (similar to unix) of make each plugin do one thing well.